### PR TITLE
Contiguous scope

### LINKED ISSUE #1835: Add contiguous scope modifier
Could be used for example targeting multiple contiguous line comments.

`"comment slurp comments fine"`

```ts
// foo
// bar
```

## Potential solution
I was thinking that this could be implemented around the existing scope handler similar to how there every modifier works and just join scopes/targets based on leading and trailing delimiter ranges.

1. Get all scopes
```ts
const scopes = scopeHandler.generateScopes(...);
```

2. Iterate scopes and join contiguous targets
```ts
const leadingRange = leadingTarget.trailingDelimiterRange ?? leadingTarget.contentRange;
const trailingRange = trailingTarget.leadingDelimiterRange ?? trailingTarget.contentRange;
if (leadingRange.intersects(trailingRange) != null) {
    newContentRange = leadingTarget.contentRange.union(trailingTarget.contentRange);
}
```

## Thinks to discuss
* One problem is that most leading and trailing delimited ranges doesn't contain the eol. eg with comment the leading range would end with the previous line and the trailing range would start with the next line. We could of course say that that is valid for merging, but it would probably make `"take slurp tokens"` join all tokens in the file and therefore take the entire file.
* Do we want to support `"every slurp <scope>"`?

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Fixes #1835 

I decided to implement this as a scope type instead of a modifier. Now you can say things like `chuck fat comment` or `changed next fat alpha`. This behavior of having alternatives to existing scopes is already something we started with `BoundedNonWhitespaceSequenceStage` albeit in a less capacity.

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [x] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+      domain: leadingTarget.contentRange,
+      getTargets: () => [leadingTarget],
+    };
+  }
+
+  const range = leadingTarget.contentRange.union(trailingTarget.contentRange);
+  return {
+    editor: leadingTarget.editor,
+    domain: range,
+    getTargets: () => [leadingTarget.withContentRange(range)],
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:119]: why not use our new `createContinuousRangeTarget` function?

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  for (const scope of generator) {
+    for (const target of scope.getTargets(false)) {
+      if (first == null) {
+        first = target;
+      }
+
+      if (last != null) {
+        const [leadingTarget, trailingTarget] = isForward
+          ? [last, target]
+          : [target, last];
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:146]: Tempted to switch to a `{proximal, distal}` obj to remove all these ternaries and confusing anonymous lists

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      last = target;
+    }
+  }
+
+  if (first != null && last != null) {
+    yield isForward ? [first, last] : [last, first];
+  }
+}
+
+function isAdjacent(leadingTarget: Target, trailingTarget: Target): boolean {
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:163]: surprised you ignore domain here. what if their domains are adjacent?

File: packages/cursorless-engine/src/scopeProviders/ScopeInfoProvider.ts [hunk 4]
```diff
@@ -176,6 +176,7 @@ function isLanguageSpecific(scopeType: ScopeType): boolean {
       return false;
 
     case "oneOf":
+    case "contiguous":
```
Comment (pokey) [inline packages/cursorless-engine/src/scopeProviders/ScopeInfoProvider.ts:179]: can't you just call this function on the constituent scope?

**Review by pokey — COMMENTED** [review]
seems like a good implementation approach. left some comments.  also, please add some tests with "next" / "previous", and be sure to test boundaries on edges of scopes and between constituent scopes

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:137]: Because I needed the continuous range for the domain before we even create the target

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  leadingTarget: Target,
+  trailingTarget: Target,
+): TargetScope {
+  if (leadingTarget.contentRange.isRangeEqual(trailingTarget.contentRange)) {
+    return {
+      editor: leadingTarget.editor,
+      domain: leadingTarget.contentRange,
+      getTargets: () => [leadingTarget],
+    };
+  }
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:131]: Should we just make a union of the domains as well you mean?

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 3 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:181]: That is on purpose. I don't think that contiguous name should eat multiple functions just because the entire function is the domain for the name.

File: packages/cursorless-engine/src/scopeProviders/ScopeInfoProvider.ts (hunk 4 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/scopeProviders/ScopeInfoProvider.ts:178]: fixed

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 1 above)
Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:137]: Not sure I follow. Why can't you use union for domain and `createContinuousRangeTarget` for the target?

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 3 above)
Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:181]: Ok, let's leave it for now; if we find we want that behaviour we can change that in the future

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+function* generateTargetRangesInDirection(
+  scopeHandler: ScopeHandler,
+  editor: TextEditor,
+  position: Position,
+  direction: Direction,
+): Iterable<[Target, Target]> {
+  const isForward = direction === "forward";
+  let first, last: Target | undefined;
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:148]: are these in iteration order or document order? We really need to have solid conventions when we could be iterating backwards

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:137]: I meant in this case where content range and domain is this same I don't want to have that calculation in two separate places. If we do a separate calculation for domain its no problem.

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 3 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:181]: Absolutely!

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 6 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:148]: First and last is in document order. The pair is yielded in iteration order.

Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:148]: Ok I'd call them `proximal` and `distal` then. That's the convention I've used elsewhere. "first" and "last" may be expected to correspond to the modifiers of those names, which are document order

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:137]: I don't mean where content range and domain is the same between two scopes. I mean the resulting content range and domain for the new scope is the same. We're lazily constructing the targets in a callback and I needed the content range to even create the scope in the first place

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 2 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts]: Switched to proximal and distal

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+    const targetsBackward = getTargetsInDirection(
+      this.scopeHandler,
+      editor,
+      position,
+      "backward",
+    );
+
+    for (const targets of targetsBackward) {
+      if (targetsForward != null && isAdjacent(targets[1], targetsForward[0])) {
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts]: done

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts]: Anyhow I now use `createContinuousRangeTarget `

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts]: done

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+  return {
+    editor: scope1.editor,
+    domain: scope1.domain.union(scope2.domain),
+    getTargets: (isReversed) => {
+      return constructScopeRangeTarget(isReversed, scope1, scope2);
+    },
+  };
+}
+
+function* generateTargetRangesInDirection(
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:94]: I'm not convinced this function is correct. Eg corner cases like single scope not adjacent to others that's last in the file. To me it looks like if you've already yielded anything, then the final yield after the loop body won't fire.

Some proper unit tests for `ContiguousScopeHandler` would help. See https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/BaseScopeHandler.test.ts if you're not sure how to do unit tests for this kinda thing. There's a limit to how far you can get using existing scopes

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+function isAdjacent(scope1: TargetScope, scope2: TargetScope): boolean {
+  if (scope1.domain.isRangeEqual(scope2.domain)) {
+    return true;
+  }
+
+  const [startTarget, endTarget] = getTargetsInDocumentOrder(
+    scope1.getTargets(false)[0],
+    scope2.getTargets(false)[0],
+  );
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:135]: do we want to throw an error if there's more than one target? Should we have a utility function `function ensureSingleTarget(scope: Scope): Target` that checks there's only one target and returns it? That worked out well for `ensureSingleEditor` I think

File: packages/cursorless-engine/src/spokenForms/defaultSpokenFormMapCore.ts [hunk 10]
```diff
@@ -130,6 +130,7 @@ export const defaultSpokenFormMapCore: DefaultSpokenFormMapDefinition = {
     next: "next",
     forward: "forward",
     backward: "backward",
+    contiguousScope: "fat",
```
Comment (pokey) [inline packages/cursorless-engine/src/spokenForms/defaultSpokenFormMapCore.ts:133]: this should prob go in `complexScopeTypes` with `glyph`, no?

**Review by pokey — COMMENTED** [review]
Well this code simplified nicely 😄. Left a couple more comments

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 8 above)
Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:94]: Actually might make sense to hold off on unit tests for the following reasons:

1. Your code actually looks at document text, unlike the test I linked, so might be harder to mock that properly
2. On second thought, I think your code is probably correct
3. I'm actually not convinced of the utility of this contiguous scope thing 😅. The only example I saw that was actually useful was "comment", and tbh I'd be tempted to just automatically merge adjacent single-line comments, as we could target a single one using "line". And I worry users will try "fat arg" to get something like https://github.com/cursorless-dev/cursorless/issues/473, and then be surprised when it breaks if there's a comment between args. So maybe we ship this with slightly fewer tests, maybe marked private? idk

Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:94]: Update from discussion today:

- We agreed "comment" is the only use case we can come up with, and we probably just want that to be default behaviour for line comments
- If @josharian wants to fight for supporting "fat" for arbitrary scopes, then we might keep it; otherwise it dies
- We potentially could reuse this code to make contiguous be default behaviour for comment

Comment (josharian) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:94]: > If @josharian wants to fight for supporting "fat" for arbitrary scopes, then we might keep it; otherwise it diesau

looking at these test cases, i think it should die

Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:94]: Sounds good. Any opinion on making single-line comments behave this way by default? (Ie auto expand to all comments that aren't separated by an empty line)

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 9 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts]: Reused our existing and shore single target. They should really never trigger with our current implementation so am fine with using an error message that might not be perfect.

**Comment by AndreasArvidsson** [comment]
@pokey @josharian The current implementation now only affects the command scope type and there is no spoken form for it.

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    yield { proximal, distal };
+  }
+}
+
+function isAdjacent(scope1: TargetScope, scope2: TargetScope): boolean {
+  if (scope1.domain.isRangeEqual(scope2.domain)) {
+    return true;
+  }
+
+  const [startTarget, endTarget] = getTargetsInDocumentOrder(
```
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:123]: This file is now general enough to be used with most scopes. If we wanted to we could remove a bunch of code here and rely on the fact that we only use it for comments. I could go either way.

**Comment by AndreasArvidsson** [comment]
Yes unless we make this more comment specific and check for single line ranges. Ideally we would have single and multi-line as separate scopes and only use this for single line.

Or add a new predicate for contiguous scope. That is probably the best solution?

**Comment by pokey** [comment]
hmm so the predicate would just set a flag on the capture. I think that sounds reasonable?

**Comment by AndreasArvidsson** [comment]
Yes exactly.

**Comment by AndreasArvidsson** [comment]
There is no problem really with this implementation except for the fact that in javascript there is no Tree sitter difference between block and line comments. I guess that contiguous predicate could take a boolean that requires single line ranges?

File: packages/cursorless-engine/src/languages/LanguageDefinition.ts [hunk 12]
```diff
@@ -141,3 +151,15 @@ function readQueryFileAndImports(languageQueryPath: string) {
...
+ * Returns true if the given scope type should use a contiguous scope handler.
+ */
+function useContiguousScopeHandler(scopeType: ScopeType): boolean {
+  switch (scopeType.type) {
+    case "comment":
+      return true;
+    default:
+      return false;
+  }
+}
```
Comment (pokey) [inline packages/cursorless-engine/src/languages/LanguageDefinition.ts:165]: I thought you were using a query predicate for this now

File: packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts [hunk 13]
```diff
@@ -185,6 +185,23 @@ class AllowMultiple extends QueryPredicateOperator<AllowMultiple> {
   }
 }
 
+/** Indicates that it's okay for this scope to extend contiguously it through its siblings. */
+class Contiguous extends QueryPredicateOperator<Contiguous> {
+  name = "contiguous!" as const;
+  schema = z.union([z.tuple([q.node]), z.tuple([q.node, q.string])]);
+
+  run(nodeInfo: MutableQueryCapture, pattern?: string) {
```
Comment (pokey) [inline packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts:193]: Yikes. This is a slippery slope. This operator merges to completely orthogonal concepts. I fear we're sleepwalking into the same problem we had with our old typescript-based matchers, where we had 5 custom functions per language.

Maybe we could add a predicate that rejects a pattern if it doesn't match a regex, and then define the pattern twice: once with positive match, one negative?

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/scope.types.ts [hunk 14]
```diff
@@ -32,6 +32,9 @@ export interface TargetScope {
    */
   readonly domain: Range;
 
+  /** Whether this scope could expand contiguously to its siblings. */
+  readonly contiguous?: boolean;
+
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/scope.types.ts:37]: huh.  A bit strange to have this on the scope handler 🤔

**Review by pokey — COMMENTED** [review]
left a few comments

File: packages/cursorless-engine/src/languages/LanguageDefinition.ts (hunk 12 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/languages/LanguageDefinition.ts:165]: We use both. I don't know how to check it at this stage. When we actually fetch the matched scope it's to late to inject this handler. Not without a lot of rewrites at least.

File: packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts (hunk 13 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts:193]: This is still not language specific. Probably more languages that dont separate block comments in tree sitter.

That could also work, but is that really easier to read and maintain?

File: packages/cursorless-engine/src/languages/LanguageDefinition.ts (hunk 12 above)
Comment (pokey) [reply packages/cursorless-engine/src/languages/LanguageDefinition.ts:165]: Sorry I'm not following how you would use both. When do you use one vs the other?

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/scope.types.ts (hunk 14 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/scope.types.ts:37]: It's on the scope and not the handler itself.

File: packages/cursorless-engine/src/languages/LanguageDefinition.ts (hunk 12 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/languages/LanguageDefinition.ts:165]: This code decides to use the handler. The the scope.contiguous is used to determine when to merge two scopes. That way we can ignore block comments.
https://github.com/cursorless-dev/cursorless/blob/b6ac60f708834daa151b24cf4d3f86f7728c4c8c/packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts#L119

File: packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts (hunk 13 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts:193]: Because the pattern matches a single line comment? look how I use it in js.

Lets.

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/scope.types.ts (hunk 14 above)
Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/scope.types.ts:37]: yeah I guess you're right 🤔

File: packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts (hunk 13 above)
Comment (pokey) [reply packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts:193]: https://github.com/cursorless-dev/cursorless/labels/to%20discuss

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts (hunk 8 above)
Comment (pokey) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts:85]: Update from meet-up:

```scm

(
  (comment) @comment
  (#match? @comment "^//")
  (#contiguous! @comment)
)
(comment) @comment
```

we both don't love this option or the option in this PR. WE prefer something like 

```scm

(
  (comment) @comment
  ($if
    (#match? @comment "^//")
    (#contiguous! @comment)
  )
)
```

but that's a bit out of scope here 😅

File: packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts (hunk 13 above)
Comment (pokey) [reply packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts:193]: Update from meet-up:

- Remove `pattern` arg from this predicate
- use the built-in tree-sitter [`match` predicate](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/binding.js#L906), and duplicate the pattern, once with `contiguous` and once without

Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts]: done!
